### DISCUSSION TITLE
PowerShell App Deployment Toolkit 3.9

### DISCUSSION THREAD

**Discussion by seanlillis** [PSAppDeployToolkit/PSAppDeployToolkit#732] [General]
Version 3.9.0 [10/01/2023]
- AddedRefreshed UI components with updated PSAppDeployToolkit branding
- Added support for Windows Toast Notifications, instead of using a BalloonTip (only for Windows 10 and above)
- Changes to dialogs so that the default system font is used for all labels and buttons, adding a more consistnent look and feel.
- Improved documentation - new overview, features and benefits detail to enhance value proposition. Improved Function documentation and overall PDF document layout.
- Updated text and structure for Comment-Based-Help to align with PowerShell Best Practices
- Show-InstallationPrompt now has a -TopMost parameter. Fixed the function failing to display a window with -NoWait parameter specified.
- Show-InstallationWelcome will now display the countdown message if there are no processes running when CloseApps timer is enabled.
- Show-InstallationProgress now supports moving the WindowLocation dynamically for existing progress dialogs (#685). When the WindowLocation is set to BottomRight, the position is now offset so as not to overlap with toast notifications (#685)
- Show-InstallationRestartPrompt now has a -TopMost parameter. Resolved an issue where parameters were not handled correctly when running asynchronously. The titlebar now shows the toolkit icon.
- Fixed vulnerability CWE-428 ensuring quotes are added to ActiveSetup paths with spaces (#726)
- Fixed -FilterApplication logic in Remove-Applications (#665)
- Fixed help text for some functions that contained old/incorrect information
- Fixed $LocalPowerUsersGroup throwing errors on machines where it does not exist (e.g. Domain Controllers)
- Resolved an issue with Set-ActiveSetup where the function would not correctly execute the process for the current user (#644)
- Resolved an issue where Set-ActiveSetup would not correctly evaluate whether to execute while running as system (#707)
- Set-ActiveSetup now also checks Version and IsInstalled properties when attempting to run the process for current user. DisableActiveSetup now also disables ExecuteForCurrentUser. Enhanced logging.
- Execute-Process now shows a shorter error description rather than the full error information in the popup window, the full error details are available in the log.
- Changed Block-AppExecution and Unblock-AppExecution so the functions also run during Non-Interactive installs (#699)
- Fixed a bug that caused Update-Desktop/Refresh-Desktop to crash Powershell on 32bit systems
- Improved calculation of dpi scale and pixels in the $GetDisplayScaleFactor variable.
- Resolved an issue where New-Folder would not create the entire path name in some scenarios.
- CHANGE: If the -ContinueOnError parameter on The Remove-Folder function now stops the script on all errors if the -ContinueOnError paramenter is set to $false. The -DisableRecursion parameter now removes everything except non-empty folders.
- Resolved an issue where the $LocalSystemNTAccount variable was not populated.
- Resolved an issue where UNC paths were not handled by New-Shortcut, Set-Shortcut and Get-Shortcut. Minor improvement to $iconIndex logic in New-Shortcut.
- Resolved an issues where the banner file object is locked and cannot be deleted in the same PowerShell session (#717)
- Updated Get-InstalledApplication to make removal of control characters less aggressive (#691)
- Updated entire toolkit to align with toolkit styling and formatting standards
- Updated Execute-MSP functions to utilize toolkit MSI database helper functions
- Added "-Force" and "-ErrorAction" parameters to "Get-ChildItem" calls in "Copy-File" function
- Added detection of defaultuser0 OOBE or ESP process to enable Non-Interactive deployment mode for Autopilot/ESP scenarios.
- Added Greek translations
- Added the appropriate exit code for MSIs that are already installed to the Execute-MSI function (#686)
- Improved Unblock-AppExecution function removal of the BlockAppExecution Schedule Task XML file
- Updated Set-PinnedApplication to use "OrdinalIgnoreCase" instead of "CurrentCultureIgnoreCase" to ignore case of pin action on Windows 10
- Set the ErrorAction of the "Unregister-Event'" command in the Finally block of the Execute-Process function to "SilentlyContinue"
- Reverted some changes in "Show-InstallationWelcome" and "Show-WelcomePrompt" to maintain compatibility with PowerShell 2.0
- Reverted changes that removed "Select-Object -ExpandProperty" calls to maintain compatibility with PowerShell 2.0
- Reverted unnecessary conversions of Where-Object pipeline pattern to less readable ForEach-Object pattern. Where-Object pattern is not more performant than ForEach-Object (ForEach or For would be, but the micro-optimization is unnecessary for our usecase).
- Reverted Resolve-Error function code to that from v3.8.3 to undo all changes

<hr /><em>This discussion was created from the release <a href='https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/releases/tag/3.9'>PowerShell App Deployment Toolkit 3.9</a>.</em>

**Comment by BP-Labs** [comment]
thankyou! super excited to start working with this new version; you have been busy!

**Comment by ClairmontAle** [comment]
Thank you for this release, small non-critical changes:
In AppDeployToolkitMain.ps1:
Line 12289, "..PARAMETER" should be ".PARAMETER"
Line 15091, "onnected" should be "Connected"
Line 15181, "Returns ca ustom" should be "Returns a custom"

**Comment by seanlillis** [comment]
Thanks for reporting these - will be fixed in the next release!

On Wed, 11 Jan 2023 at 22:52, ClairmontAle ***@***.***> wrote:

> Thank you for this release, small non-critical changes:
> In AppDeployToolkitMain.ps1:
> Line 12289, "..PARAMETER" should be ".PARAMETER"
> Line 15091, "onnected" should be "Connected"
> Line 15181, "Returns ca ustom" should be "Returns a custom"
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/discussions/732#discussioncomment-4661615>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC3A6MJGX6U4PPZJUZ652BDWR42SJANCNFSM6AAAAAATXLWVUQ>
> .
> You are receiving this because you authored the thread.Message ID:
> <PSAppDeployToolkit/PSAppDeployToolkit/repo-discussions/732/comments/4661615
> @github.com>
>


-- 
Seán

**Comment by s016aq** [comment]
Is it only me who can't use this at all? It only works if I'm logged in as the same user that is executing DeployApplication.exe.
I.e. if I'm logged in as a user and executing as SYSTEM (or administrator) it crashes:
<img width="197" alt="psadt" src="https://user-images.githubusercontent.com/122624425/212351729-b98dc8c2-2183-4680-98b8-6c944e77461c.png">
This has of course never been a problem with previous versions. Been using this since version 1.0.0
Any ideas?!

**Reply by RichardLafontaine** [reply]
it's this issue: #733

**Comment by rvkasper** [comment]
@seanlillis @sintaxasn I can attempt a pr. Basically copy how the asynchronous balloon tip passes a script block, only for the toast if the processntaccount<>runasactiveuser.ntaccount pass script block for the toast to execute-processasuser? If I get it working I'll submit the PR this weekend

**Comment by freedbygrace** [comment]
Guys thank you very much! I have several functions I believe would add tremendous value to the toolkit. How can I contribute them?!

**Comment by ocsscott** [comment]
There is a forum for extensions.  Out of interest what are they?

Sent from my T-Mobile 5G Device
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by freedbygrace** [reply]
Invoke-CommandLineParser - This could take a series of uninstall strings and reliably break them into command and arguments. Would help greatly with dynamic application removal.

Arguments may also be added/removed as need then used with Execute-Process.

I have several more that would add great utility.

**Reply by That-Annoying-Guy** [reply]
Please post this and all the functions you want to share at https://discourse.psappdeploytoolkit.com/c/thetoolkit/extensions/13

One post for each function. 
If possible, describe how to use it either in the code or in the post.

**Comment by ocsscott** [comment]
I'd suggest  posting in extensions id love to see what you created

Sent from my T-Mobile 5G Device
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by CKurti-MCMTSG** [comment]
May I suggest that your release tag be updated from "3.9" to "3.9.0" so that release tags are consistant with previous ones as well as with the file name?

**Reply by seanlillis** [reply]
This is done, thanks.
